### PR TITLE
Skip password auth if client cert required (#230)

### PR DESCRIPTION
**Pull request by sriramy**

### CODE REVIEW

**Comment by drakkan** [comment]
Hi,

I don't like this approach, to be able to reuse this for a future SFTPGo version I would like to handle the logic outside ftpserverlib, please take a look at #232. I added an optional interface:

```
VerifyConnection(cc ClientContext, user string, tlsConn *tls.Conn) (ClientDriver, error)
```

 if you implement it you can check the certificate there. If this interface returns an error the client will be disconnected, if it returns no error and nil driver the password is required, if it returns nil error and not nil driver the client is authenticated.

This way the certificate verification logic will be handled in this callback implemented inside your application.

I haven't tested my patch with any real client, please give it a try, thank you

File: handle_auth.go [hunk 1]
```diff
@@ -10,6 +12,16 @@ func (c *clientHandler) handleUSER(param string) error {
 		return nil
 	}
 
+	if c.HasTLSForControl() && c.server.settings.SkipPasswordIfClientCertMatchesUser {
+		clientCertificates := c.controlTLSConn.ConnectionState().PeerCertificates
+		// TODO: Add check for subjectAltName as well
+		if len(clientCertificates) > 0 && clientCertificates[0].Subject.CommonName == param {
+			c.user = param
+			c.writeMessage(StatusUserLoggedIn, "User logged in, no password required.")
```
Comment (drakkan) [inline handle_auth.go:20]: You don't set a driver here, I haven't tested this patch but I think that without a driver the client will be unable to do anything

Comment (sriramy) [reply handle_auth.go:20]: I agree, I didn't much testing before creating a pull request. Sorry about that. Tested your patch though, and it seems to work for me.

**Comment by sriramy** [comment]
Abandoned in favor of https://github.com/fclairamb/ftpserverlib/pull/232

**Pull request closed without merging**